File: treeSort.js
//Binary search tree - From: https://www.geeksforgeeks.org/implementation-binary-search-tree-javascript/

class Node {
    constructor(data) {
        this.key = getDistanceFromLatLonInKm(data.lat, data.lon);
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
        this.sorted = [];
    }

    getSorted() {
        return this.sorted;
    }

    insert(data) {
        const newNode = new Node(data);

        if (this.root === null) {
            this.root = newNode;
        } else {
            this.insertNode(this.root, newNode);
        }
    }

    insertNode(node, newNode) {
        if (newNode.key < node.key) {
            if (node.left === null) {
                node.left = newNode;
            } else {
                this.insertNode(node.left, newNode);
            }
        } else {
            if (node.right === null) {
                node.right = newNode;
            } else {
                this.insertNode(node.right, newNode);
            }
        }
    }

    remove(node) {
        this.root = this.removeNode(this.root, node);
    }

    removeNode(root, node) {
        if (root === null) {
            return null;
        } else if (node.key < root.key) {
            root.left = this.removeNode(root.left, node);
            return root;
        } else if (node.key > root.key) {
            root.right = this.removeNode(root.right, node);
            return root;
        } else {
            if (root.left === null && root.right === null) {
                root = null;
                return root;
            }
            if (root.left === null) {
                root = root.right;
                return root;
            } else if (root.right === null) {
                root = root.left;
                return root;
            }

            const aux = this.findMinNode(root.right);
            root.data = aux.data;

            root.right = this.removeNode(root.right, aux.data);
            return root;
        }
    }

    findMinNode(node) {
        if (node.left === null) {
            return node;
        } else {
            return this.findMinNode(node.left);
        }
    }

    inorder(node) {
        if (node !== null) {
            this.inorder(node.left);
            this.sorted.push(node.data);
            const right = node.right;
            this.remove(node);
            this.inorder(right);
        }
    }
}

File: aspnet/Examen/TreeSort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

//Binary search tree - From: https://www.geeksforgeeks.org/implementation-binary-search-tree-javascript/ - Ported to C# by me

namespace Examen
{
    public class Node
    {
        Haversine haversine = new Haversine();
        public GPSData data;
        public Node left;
        public Node right;
        public double key;

        public Node(GPSData data)
        {
            this.data = data;
            this.left = null;
            this.right = null;
            this.key = haversine.getDistance(data.lat, data.lon);
        }
    }

    public class BinarySearchTree
    {
        public Node root;
        public List<GPSData> sorted;

        public BinarySearchTree()
        {
            root = null;
            sorted = new List<GPSData>();
        }

        public void Insert(GPSData data)
        {
            Node newNode = new Node(data);

            if (root == null)
            {
                root = newNode;
            }
            else
            {
                InsertNode(root, newNode);
            }
        }

        private void InsertNode(Node node, Node newNode)
        {
            if (newNode.key < node.key)
            {
                if (node.left == null)
                {
                    node.left = newNode;
                }
                else
                {
                    InsertNode(node.left, newNode);
                }
            }
            else
            {
                if (node.right == null)
                {
                    node.right = newNode;
                }
                else
                {
                    InsertNode(node.right, newNode);
                }
            }
        }

        private void Remove(Node node)
        {
            root = RemoveNode(root, node);
        }

        private Node RemoveNode(Node root, Node node)
        {
            if (root == null)
            {
                return null;
            }
            else if (node.key < root.key)
            {
                root.left = RemoveNode(root.left, node);
                return root;
            }
            else if (node.key > root.key)
            {
                root.right = RemoveNode(root.right, node);
                return root;
            }
            else
            {
                if (root.left == null && root.right == null)
                {
                    root = null;
                    return root;
                }
                if (root.left == null)
                {
                    root = root.right;
                    return root;
                }
                else if (root.right == null)
                {
                    root = root.left;
                    return root;
                }

                Node aux = FindMinNode(root.right);
                root.data = aux.data;

                root.right = RemoveNode(root.right, aux);
                return root;
            }
        }

        private Node FindMinNode(Node node)
        {
            if (node.left == null)
            {
                return node;
            }
            else
            {
                return FindMinNode(node.left);
            }
        }

        public void Inorder(Node node)
        {
            if (node != null)
            {
                Inorder(node.left);
                sorted.Add(node.data);
                Node right = node.right;
                Remove(node);
                Inorder(right);
            }
        }
    }
}
